src/java/main/org/apache/zookeeper/server/quorum/CommitProcessor.java

```diff
@@ -240,84 +240,14 @@ public void run() {
-                             * request), the finalRequestProcessor will see a closed cnxn handle, and just won't send a
-                             * response.
-                             * Also note that we don't have a local session, therefore we treat the request
-                             * like any other commit for a remote request, i.e., we perform the update without sending
-                             * a response.
-                             */
-                            LOG.warn("Got request " + request +
-                                    " but we are expecting request " + topPending);
-                            sessionQueue.addFirst(topPending);
-                        } else {
-                            /*
-                             * Generally, we want to send to the next processor our version of the request,
-                             * since it contains the session information that is needed for post update processing.
-                             * In more details, when a request is in the local queue, there is (or could be) a client
-                             * attached to this server waiting for a response, and there is other bookkeeping of
-                             * requests that are outstanding and have originated from this server
-                             * (e.g., for setting the max outstanding requests) - we need to update this info when an
-                             * outstanding request completes. Note that in the other case (above), the operation
-                             * originated from a different server and there is no local bookkeeping or a local client
-                             * session that needs to be notified.
-                             */
-                            topPending.setHdr(request.getHdr());
-                            topPending.setTxn(request.getTxn());
-                            topPending.zxid = request.zxid;
-                            request = topPending;
-                        }
-                    }
-
-                    sendToNextProcessor(request);
-
-                    waitForEmptyPool();
-
-                    /*
-                     * Process following reads if any, remove session queue if
-                     * empty.
-                     */
-                    if (sessionQueue != null) {
-                        while (!stopped && !sessionQueue.isEmpty()
-                                && !needCommit(sessionQueue.peek())) {
-                            sendToNextProcessor(sessionQueue.poll());
-                        }
-                        // Remove empty queues
-                        if (sessionQueue.isEmpty()) {
-                            pendingRequests.remove(request.sessionId);
+                    // We only need to perform synchronization if we are on the last request in the queue
+                    if (committedRequests.size() == 1) {
+                        synchronized (committedRequests) {
```

Comment:
Unfortunately I don't get why do you need to synchronise here. Would you please elaborate a little bit?